web_src/js/features/comp/ImagePaste.js

```diff
@@ -1,94 +1,72 @@
-
-  const files = [];
-  for (const item of e.clipboardData.items || []) {
-    if (!item.type || !item.type.startsWith('image/')) continue;
-    files.push(item.getAsFile());
-  }
-  return files;
-}
-
-class TextareaEditor {
-  constructor(editor) {
-    this.editor = editor;
-  }
-
-  insertPlaceholder(value) {
-    const editor = this.editor;
-    const startPos = editor.selectionStart;
-    const endPos = editor.selectionEnd;
-    editor.value = editor.value.substring(0, startPos) + value + editor.value.substring(endPos);
-    editor.selectionStart = startPos;
-    editor.selectionEnd = startPos + value.length;
-    editor.focus();
+import {getAttachedEasyMDE} from './EasyMDE.js';
+
+/**
+ *
+ * @param {*} editor
+ * @param {*} file
+ */
+export function addUploadedFileToEditor(editor, file) {
+  if (!editor && file.previewElement) {
+    editor = file.previewElement.closest('form') ? getAttachedEasyMDE(file.previewElement.closest('form').querySelector('textarea')) : getAttachedEasyMDE(file.previewElement.parentElement.parentElement.parentElement.querySelector('textarea'));
+    editor = editor.codemirror;
   }
-
-  replacePlaceholder(oldVal, newVal) {
-    const editor = this.editor;
-    const startPos = editor.selectionStart;
-    const endPos = editor.selectionEnd;
-    if (editor.value.substring(startPos, endPos) === oldVal) {
-      editor.value = editor.value.substring(0, startPos) + newVal + editor.value.substring(endPos);
-      editor.selectionEnd = startPos + newVal.length;
+  const startPos = editor.selectionStart || editor.getCursor && editor.getCursor('start');
+  const endPos = editor.selectionEnd || editor.getCursor && editor.getCursor('end');
+  const isimage = file.type.startsWith('image/') ? '!' : '';
+  const fileName = (isimage ? file.name.replace(/\.[^/.]+$/, '') : file.name);
+  if (startPos) {
+    if (editor.setSelection) {
+      if (startPos.line !== undefined) {
+        editor.setSelection(startPos, endPos);
+        editor.replaceSelection(`${isimage}[${fileName}](/attachments/${file.uuid})\n`);
+      } else {
+        const val = editor.getValue();
+        editor.setValue(`${val}\n${isimage}[${fileName}](/attachments/${file.uuid})`);
+      }
+    } else if (typeof startPos === 'number' && startPos > 0) {
+      editor.value = `${editor.value.substring(0, startPos)}\n${isimage}[${fileName}](/attachments/${file.uuid})\n${editor.value.substring(endPos)}`;
     } else {
-      editor.value = editor.value.replace(oldVal, newVal);
-      editor.selectionEnd -= oldVal.length;
-      editor.selectionEnd += newVal.length;
+      editor.value += `\n${isimage}[${fileName}](/attachments/${file.uuid})`;
     }
-    editor.selectionStart = editor.selectionEnd;
-    editor.focus();
+  } else if (editor.setSelection) {
+    editor.value(`${editor.value()}\n${isimage}[${fileName}](/attachments/${file.uuid})\n`);
+  } else {
+    editor.value += `${editor.value}\n${isimage}[${fileName}](/attachments/${file.uuid})\n`;
```

Comment:
Is there a logic behind the final `\n`? I'd generally suggest a helper function to create these strings instead and that can take a boolean whether it should add the newline or not.

Reply:
The old code has an abstraction layer for different editors (TextareaEditor/CodeMirrorEditor) to do the placeholder inserting/replacing, I think the old code is clearer and more maintainable.

And I still think the placeholder is necessary for users to get obvious feedbacks in all cases.

Reply:
I to refactor again and add placeholder, please review again
Thanks